simple_certmanager/admin.py

```diff
@@ -89,16 +88,66 @@ class SigningRequestAdmin(admin.ModelAdmin):
         ),
     )
 
+    def get_object(
+        self, request: HttpRequest, object_id: str, from_field: str | None = None
+    ):
+        """
+        Here I only access the object so I can check which message to display
+        to help the user. I don't need to change the object in any way.
+        """
+        object = super().get_object(request, object_id, from_field)
+        if object and object.public_certificate:
+            messages.warning(
+                request,
+                _(
+                    "This request is signed therefore can not be edited."
+                    " You can safely delete it."
+                ),
+            )
+
+        return object
+
+    def get_readonly_fields(
+        self, request: HttpRequest, obj: SigningRequest | None = None
+    ):
+        "Make the CSR field read-only after it has been signed."
+        readonly_fields = [
+            "csr",
+            "public_certificate",
+        ]
+        if obj and obj.public_certificate:
+            readonly_fields.extend(
+                [
+                    "common_name",
+                    "organization_name",
+                    "country_name",
+                    "state_or_province_name",
+                    "locality_name",
+                    "email_address",
+                ]
+            )
+        return readonly_fields
```

Comment:
I think you can simplify this a lot by overriding the method `has_change_permission` instead.

The `messages` approach may also not be the best since you can trigger the view which will put the message in the session, but it may then be displayed on another page which would be confusing. It's a bit of a race condition, but I've seen cases where messages are displayed on the login screen *after* logging out of the admin.

Reply:
OK for the permission.

Do you see another way to notify the users why their permission has changed though ? We looked for ways to override the default displayed message unsuccessfully. 
 
IMHO, it is preferable to have some information that gives notice to the user that he cannot edit an instance anymore rather than no notice but I also agree that the messages might display when not desired.

Reply:
hrm, I thought this mechanism was viable to automatically make the thing readonly (change vs. view permission), but from your comment I have the impression that is not the case?

Reply:
Yes the permission mechanism DOES make the instance read-only.

My issue is in the fact that the instance "magically" switches from editable to read-only. 
As developer, we know why it switched but the user might be clueless about why his instance became read-only and that is what I tried to address with the messages/ get_object method.

Reply:
check - I think it's fine to leave it as is. It's the admin environment, which should not be exposed to people that are completely clueless. Keeping in mind the people who we are building this feature for, I think it's clear enough and we can still document this on read the docs if people ask questions.

It's Good Enough.

Reply:
allright. Removed the code and added the relevant doc.